crypto/asn1/i2d_evp.c

```diff
@@ -25,29 +25,42 @@
 #include "crypto/asn1.h"
 #include "crypto/evp.h"
 
+struct type_and_structure_st {
+    const char *output_type;
+    const char *output_structure;
+};
+
 static int i2d_provided(const EVP_PKEY *a, int selection,
-                        const char *output_structures[],
+                        const struct type_and_structure_st *output_info,
                         unsigned char **pp)
 {
     OSSL_ENCODER_CTX *ctx = NULL;
     int ret;
 
     for (ret = -1;
-         ret == -1 && *output_structures != NULL;
-         output_structures++) {
+         ret == -1 && output_info->output_type != NULL;
+         output_info++) {
         /*
          * The i2d_ calls don't take a boundary length for *pp.  However,
-         * OSSL_ENCODER_CTX_get_num_encoders() needs one, so we make one
-         * up.
+         * OSSL_ENCODER_to_data() needs one, so we make one up.  Because
+         * OSSL_ENCODER_to_data() decrements this number by the amount of
+         * bytes written, we need to calculate the length written further
+         * down, when pp != NULL.
          */
         size_t len = INT_MAX;
 
-        ctx = OSSL_ENCODER_CTX_new_for_pkey(a, selection, "DER",
-                                            *output_structures, NULL);
+        ctx = OSSL_ENCODER_CTX_new_for_pkey(a, selection,
+                                            output_info->output_type,
+                                            output_info->output_structure,
+                                            NULL);
         if (ctx == NULL)
             return -1;
-        if (OSSL_ENCODER_to_data(ctx, pp, &len))
-            ret = (int)len;
+        if (OSSL_ENCODER_to_data(ctx, pp, &len)) {
+            if (pp == NULL)
+                ret = (int)len;
+            else
+                ret = INT_MAX - (int)len;
```

Comment:
Should we add a comment here referring to the earlier comment (i.e., that this is where we "calculate the length written")?